### PR TITLE
[core][autoscaler][v2] scheduler - resource requests impl [3/x]

### PR DESCRIPTION
**Pull request by rickyyx**
## Why are these changes needed?
See https://github.com/ray-project/ray/pull/39851 for a draft prototype.

This PR adds the code needed for scheduling resource requests (tasks, actors).

In a high level it does the following: for each resource request, it tries to find an existing or new node to fulfill the request.
- When finding the node, it uses a greedy approach to try putting the request on each available node (or new node) and rank them based on a utlization score. 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/autoscaler/v2/scheduler.py [hunk 1]
```diff
@@ -100,11 +99,24 @@ class SchedulingNodeStatus(Enum):
...
+    RAY_PENDING = "PENDING"
+    # ray is already running.
     RUNNING = "RUNNING"
 
+    @staticmethod
+    def is_ray_pending(instance: Instance) -> bool:
+        """
+        Check if the instance is pending on ray.
+        """
+        return instance.status in [
```
Comment (jjyao) [inline python/ray/autoscaler/v2/scheduler.py:112]: I remember InstanceUtil has the same function?

File: python/ray/autoscaler/v2/scheduler.py [hunk 2]
```diff
@@ -209,7 +246,74 @@ def _compute_score(self) -> UtilizationScore:
...
+        # TODO: add labels when implementing the gang scheduling.
+
+        # Check if there's enough resources to schedule the request.
+        for k, v in request.resources_bundle.items():
+            if self.available_resources.get(k, 0) < v:
+                return False
+
+        # Schedule the request, update resources
+        for k, v in request.resources_bundle.items():
+            self.available_resources[k] -= v
```
Comment (jjyao) [inline python/ray/autoscaler/v2/scheduler.py:311]: Do you handle the implicit resources here?

File: python/ray/autoscaler/v2/scheduler.py [hunk 3]
```diff
@@ -209,7 +246,74 @@ def _compute_score(self) -> UtilizationScore:
         Returns:
             A utilization score for this node.
         """
-        pass
+
+        # Compute the number of resource types being scheduled.
+        num_matching_resource_types = 0
+        for req in self.sched_requests:
+            for resource_name in req.resources_bundle.keys():
+                if resource_name in self.total_resources:
```
Comment (jjyao) [inline python/ray/autoscaler/v2/scheduler.py:254]: I think it will never be false?

File: python/ray/autoscaler/v2/scheduler.py [hunk 4]
```diff
@@ -175,7 +187,32 @@ def try_schedule(
                 - the utilization score for this node with respect to the current
                 resource requests being scheduled.
         """
-        pass
+        # Track the resource requests that cannot be scheduled on this node.
+        unschedulable_requests = []
+
+        def _sort_resource_request(req: ResourceRequest) -> Tuple:
+            """
+            Sort the resource requests by:
```
Comment (jjyao) [inline python/ray/autoscaler/v2/scheduler.py:195]: why this particular sort order?

File: python/ray/autoscaler/v2/scheduler.py (hunk 1 above)
Comment (rickyyx) [reply python/ray/autoscaler/v2/scheduler.py:112]: right - modified.

File: python/ray/autoscaler/v2/scheduler.py (hunk 2 above)
Comment (rickyyx) [reply python/ray/autoscaler/v2/scheduler.py:311]: yeah, these are directly taken from node states, which are passed through from node resources.

File: python/ray/autoscaler/v2/scheduler.py (hunk 3 above)
Comment (rickyyx) [reply python/ray/autoscaler/v2/scheduler.py:254]: why so? could there be a resource bundle e.g. ('some-unknown-resource') that's not present in the total resources of a node?

File: python/ray/autoscaler/v2/scheduler.py (hunk 4 above)
Comment (rickyyx) [reply python/ray/autoscaler/v2/scheduler.py:195]: We had the similar logic in the existing implementation (other than having placement constraints):

https://github.com/ray-project/ray/blob/c21f7a209a0ccbadd3bdf73eb3bd6240fabc629c/python/ray/autoscaler/_private/resource_demand_scheduler.py#L902-L914

**Comment by rickyyx** [comment]
Moved to https://github.com/ray-project/ray/compare/pr-auto-sched-resource-reqs?expand=1 for stacked PR.

**Pull request closed without merging**